src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java

```diff
@@ -212,10 +213,33 @@ private String buildUrl(String impid, ExtImpAdocean extImpAdocean, String consen
             uriBuilder.addParameter("gdpr", "1");
         }
 
+        final User user = bidRequest.getUser();
         if (user != null && StringUtils.isNotEmpty(user.getBuyeruid())) {
             uriBuilder.addParameter("hcuserid", user.getBuyeruid());
         }
 
+        final App app = bidRequest.getApp();
+        if (app != null) {
+            uriBuilder.addParameter("app", "1");
+            uriBuilder.addParameter("appname", app.getName());
+            uriBuilder.addParameter("appbundle", app.getBundle());
+            uriBuilder.addParameter("appdomain", app.getDomain());
+        }
+
+        final Device device = bidRequest.getDevice();
+        if (device != null) {
+            if (StringUtils.isNotEmpty(device.getIfa())) {
+                uriBuilder.addParameter("ifa", device.getIfa());
+            } else {
+                uriBuilder.addParameter("dpidmd5", device.getDidmd5());
+            }
+
+            uriBuilder.addParameter("devos", device.getOs());
+            uriBuilder.addParameter("devosv", device.getOsv());
+            uriBuilder.addParameter("devmodel", device.getModel());
+            uriBuilder.addParameter("devmake", device.getMake());
```

Comment:
Seems we should add this values to integration test auction-request, and also create Unit tests, specially for this case